src/components/Cluster/ClusterDetail/EditClusterLabelsModal/EditClusterLabelsModal.tsx

```diff
@@ -0,0 +1,173 @@
+import * as clusterLabelsActions from 'actions/clusterLabelsActions';
+import PropTypes from 'prop-types';
+import React, {
+  Reducer,
+  useCallback,
+  useMemo,
+  useReducer,
+  useState,
+} from 'react';
+import { connect, DispatchProp } from 'react-redux';
+import { bindActionCreators, Dispatch } from 'redux';
+import Button from 'UI/Button';
+import ClusterIDLabel from 'UI/ClusterIDLabel';
+import EditableValueLabel from 'UI/EditableValueLabel';
+
+import GenericModal from '../../../Modals/GenericModal';
+import AddClusterLabel from './AddClusterLabel';
+
+interface ILabels {
+  [key: string]: string;
+}
+
+interface IStateProps {
+  loading: boolean;
+}
+
+interface IDispatchProps extends DispatchProp {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  actions: Record<string, (...args: any[]) => Promise<any>>;
+}
+
+interface IEditClusterLabelsModalProps extends IDispatchProps, IStateProps {
+  clusterId: string;
+  labels: ILabels;
+}
+
+interface IUpdateLabelReducerAction {
+  label: string;
+  value: string;
+  reset?: boolean;
+}
+
+const EditClusterLabelsModal = ({
+  clusterId,
+  labels,
+  actions,
+  loading,
+}: IEditClusterLabelsModalProps) => {
+  const editableLabels = useMemo(() => {
+    const editable: ILabels = {};
+
+    for (const labelKey of Object.keys(labels)) {
+      if (labelKey.includes('giantswarm.io') === false) {
+        editable[labelKey] = labels[labelKey];
+      }
+    }
+
+    return editable;
+  }, [labels]);
+
+  const [modifiedLabels, updateLabel] = useReducer<
+    Reducer<ILabels, IUpdateLabelReducerAction>
+  >((oldLabels, newLabel) => {
+    if (newLabel.reset) {
+      return {};
+    }
+
+    return {
+      ...oldLabels,
+      [newLabel.label]: newLabel.value,
+    };
+  }, {});
+
+  const displayLabels = useMemo(
+    () => ({ ...editableLabels, ...modifiedLabels }),
+    [editableLabels, modifiedLabels]
+  );
+
+  const [open, setOpen] = useState(false);
+  const closeModal = () => {
+    setOpen(false);
+    // supply label & value here, typescript demands it
+    updateLabel({ label: '', value: '', reset: true });
+  };
+
+  const saveLabels = useCallback(() => {
+    actions.clusterLabelsPatch(clusterId, modifiedLabels).then(() => {
+      closeModal();
+    });
+  }, [actions, clusterId, modifiedLabels]);
+
+  return open ? (
```

Comment:
Can we use our new "return early" pattern here? (ex: if not open, return the edit button, and then the rest)